File: 111/111.py
# Online Judge : UVA Online Judge
# Problem No   : 111
# Problem Name : History Grading
# Author       : Zubayer Rahman
# Email        : zubayer.csesust@gmail.com
# Time Limit   : 3.000s
# CPU          : 0.130s
# Memory       :
# Algorithm    : LCS

import sys
from typing import List


def LCS(s1: List[int], s2: List[int]) -> int:
    lcs = [[0 for _ in range(max(len(s1), len(s2)) + 1)] for _ in range(2)]

    bi = bool

    for i in range(len(s1) + 1):
        bi = i & 1
        for k in range(len(s2) + 1):
            if i == 0 or k == 0:
                lcs[bi][k] = 0
            elif s1[i - 1] == s2[k - 1]:
                lcs[bi][k] = 1 + lcs[1 - bi][k - 1]
            else:
                lcs[bi][k] = max(lcs[1 - bi][k], lcs[bi][k - 1])

    return lcs[bi][len(s2)]


def main():
    try:
        sys.stdin = open("input", "r")
    except:
        pass

    while True:
        n = int(sys.stdin.readline().strip())
        order = sys.stdin.readline().strip().split(" ")

        s1 = [0 for _ in range(21)]
        for idx, num in enumerate(order):
            s1[int(num)] = idx + 1

        while True:
            line = sys.stdin.readline().strip()
            if not line:
                break

            line = line.split(" ")
            s2 = [0 for _ in range(21)]
            for idx, num in enumerate(line):
                s2[int(num)] = idx + 1

            s1 = [x for x in s1 if x]
            s2 = [x for x in s2 if x]

            score = LCS(s1, s2)
            sys.stdout.write(str(score) + "\n")

        break


if __name__ == '__main__':
    main()

File: 111/111.cpp
/**
Online Judge : UVA Online Judge
Problem No   : 111
Problem Name : History Grading
Author       : Zubayer Rahman
Email        : zubayer.csesust@gmail.com
Time Limit   : 3.000s
CPU          : 0.000s
Memory       :
Algorithm    : LCS
*/

#include<bits/stdc++.h>

using namespace std;

int LCS(int s1[], int s2[], int length)
{
    int lcs [2][length+1] = {{0}};

    for(int i=0; i<2; i++)
        for(int k=0; k<length+1; k++)
            lcs[i][k] = 0;

    bool bi;

    for(int i=0; i<=length+1; i++)
    {
        bi = i & 1;
        for(int k=0; k<=length+1; k++)
        {
            if(i==0 || k==0)
                lcs[bi][k+1] = 0;
            else if(s1[i] == s2[k])
                lcs[bi][k+1] = 1 + lcs[1-bi][k+1];
            else lcs[bi][k+1] = max(lcs[1-bi][k+1], lcs[bi][k]);
        }
    }

    return lcs[bi][length+1];
}

int main()
{
#ifndef ONLINE_JUDGE
    freopen("input", "r", stdin);
#endif // ONLINE_JUDGE

    int n;

    while(scanf("%d", &n) == 1)
    {
        int s1[n+1], s2[n+1];

        for (int i=1; i<=n; i++)
        {
            int a;

            scanf("%d", &a);
            s1[a] = i;
        }

        while(true)
        {
            int a;

            for(int i=1; i<=n; i++)
            {
                if (scanf("%d", &a) != 1)
                    return 0;

                s2[a] = i;
            }

            int result = LCS(s1, s2, n);
            printf("%d\n", result);
        }
    }

    return 0;
}
